lib/services/address/index.js

```diff
@@ -224,55 +303,71 @@ AddressService.prototype.updateMempoolIndex = function(tx) {
+    var inputOutputIndexBuffer = new Buffer(4);
+    inputOutputIndexBuffer.writeUInt32BE(input.outputIndex);
+
+    // Add an additional small spent index for fast synchronous lookups
+    var spentIndexSyncKey = this._encodeSpentIndexSyncKey(
+      input.prevTxId,
+      input.outputIndex
+    );
+    if (add) {
+      this.mempoolSpentIndex[spentIndexSyncKey] = true;
+    } else {
+      delete this.mempoolSpentIndex[spentIndexSyncKey];
+    }
 
+    // Add a more detailed spent index with values
+    var spentIndexKey = Buffer.concat([
+      AddressService.MEMPREFIXES.SPENTSMAP,
+      input.prevTxId,
+      inputOutputIndexBuffer
+    ]);
     var inputIndexBuffer = new Buffer(4);
     inputIndexBuffer.writeUInt32BE(inputIndex);
     var inputIndexValue = Buffer.concat([
       txidBuffer,
       inputIndexBuffer
     ]);
-    this.mempoolSpentIndex[spentIndexKey] = inputIndexValue;
+    operations.push({
+      type: action,
+      key: spentIndexKey,
+      value: inputIndexValue
+    });
 
-    var address = input.script.toAddress(this.node.network);
-    if (!address) {
+    // Update input index
+    var inputHashBuffer;
+    if (input.script.isPublicKeyHashIn()) {
+      inputHashBuffer = Hash.sha256ripemd160(input.script.chunks[1].buf);
+    } else if (input.script.isScriptHashIn()) {
+      inputHashBuffer = Hash.sha256ripemd160(input.script.chunks[input.script.chunks.length - 1].buf);
+    } else {
       continue;
     }
-    var addressStr = address.toString();
-    if (!this.mempoolInputIndex[addressStr]) {
-      this.mempoolInputIndex[addressStr] = [];
-    }
-    this.mempoolInputIndex[addressStr].push({
-      txid: tx.hash, // TODO use buffer
-      inputIndex: inputIndex
+    var inputKey = Buffer.concat([
+      AddressService.MEMPREFIXES.SPENTS,
+      inputHashBuffer,
+      input.prevTxId,
+      inputOutputIndexBuffer
+    ]);
+    var inputValue = Buffer.concat([
+      txidBuffer,
+      inputIndexBuffer
+    ]);
+    operations.push({
+      type: action,
+      key: inputKey,
+      value: inputValue
     });
-  }
 
-};
+  }
 
-/**
- * This function is called by the Database Service when the database itself
- * has finished synchronization. It will retrieve a copy of all transactions
- * from the mempool and create an address index for fast look-ups. The previous
- * index will be reset.
- */
-AddressService.prototype.resetMempoolIndex = function(callback) {
-  var self = this;
-  var transactionBuffers = self.node.services.bitcoind.getMempoolTransactions();
-  this.mempoolInputIndex = {};
-  this.mempoolOutputIndex = {};
-  this.mempoolSpentIndex = {};
-  async.each(transactionBuffers, function(txBuffer, next) {
-    var tx = Transaction().fromBuffer(txBuffer);
-    self.updateMempoolIndex(tx);
-    setImmediate(next);
-  }, function(err) {
+  this.mempoolIndex.batch(operations, function(err) {
     if (err) {
-      return callback(err);
+      return log.error(err);
```

Comment:
We may be able to have an optional callback, so that this block https://github.com/braydonf/bitcore-node/blob/mem/lib/services/address/index.js#L245 can be called after this action.